pkg/controller/operconfig/migration.go

```diff
@@ -32,6 +43,84 @@ func migrateEgressFirewallCRs(ctx context.Context, operConfig *operv1.Network, c
 	return nil
 }
 
+func enableMulticastOVN(ctx context.Context, client cnoclient.Client) error {
+	// 1. query for netnamespaces
+	nnsList, err := client.Default().Dynamic().Resource(gvrNetnamespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	// 2. iterate through netnamespaces
+	//    - any with multicast-enabled="true" annotation will cause an update to the corresponding
+	//      namespace to add the necessary OVN annotation
+	for _, nns := range nnsList.Items {
+		if nns.Object["metadata"].(map[string]interface{})["annotations"] == nil {
+			continue
+		}
+		multicastAnnotation := nns.Object["metadata"].(map[string]interface{})["annotations"].(map[string]interface{})["netnamespace.network.openshift.io/multicast-enabled"]
+		if multicastAnnotation == "true" {
+			// first update namespace to have the same annotation
+			nspStr := fmt.Sprintf("%v", nns.Object["netname"])
+			namespaceObj, err := client.Default().Kubernetes().CoreV1().Namespaces().Get(ctx, nspStr, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			namespaceObj.Annotations["k8s.ovn.org/multicast-enabled"] = "true"
+
+			if err := apply.ApplyObject(ctx, client, namespaceObj, ""); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func enableMulticastSDN(ctx context.Context, client cnoclient.Client) error {
+	multicastRollbackReady, err := netNamespacesExistForAllNamespaces(ctx, client)
+	if !multicastRollbackReady {
+		return nil // wait for all SDN netnamespace resources to be created before rollback
```

Comment:
It seems to me that returning `nil` will just cause it to think the migration completed successfully?

Reply:
This is acceptable behavior AFAIU, and is actually necessary for this migration rollback to work. Here is my understanding:

- No error has occurred, so we don't want to return an error. 

- See where the [function is called](https://github.com/openshift/cluster-network-operator/pull/1543/files#diff-7c8b030b905dedc903155ca81c4f4089318866fece8d863fa881ece4f46f9333L403-R412), the migration process will execute until the operConfig.Spec.Migration field becomes nil. We actually do want the migration to "complete successfully" by doing nothing and returning nil -- until the multicastRollback is ready (NetNamespaces should exist for all namespaces), at which point the migration truly will execute and complete successfully. 
- Note that this code, the egressFirewall migration code, and the egressIP migration code all will continue to execute as part of the CNO until the migration field is set to null.
- According to openshift docs the user must manually set the migration field to null by patching the CNO once certain conditions for migration rollback are met. My code is in line with this, because the NetNamespace objects will be recreated before said conditions are met, and the code beyond what you have highlighted will execute.

Reply:
Will there be any issue if we return an error here? An error message can help us to debug. Also normally, we shall return an error if the desired change cannot be done in a reconciliation cycle.

Reply:
If the err in the highlighted code is non nil, we do return that. So yes it is as Dan said.

Reply:
> Is there a Jira issue for updating the docs to mention the new criteria?

I'm not sure what you mean. The criteria has not changed.